### PR TITLE
API: Timestamp(pydatetime) microsecond reso

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

I've got one json test still failing locally cc @WillAyd see the comment in the ujson file, suggestions?

### CODE REVIEW

**Comment by WillAyd** [comment]
I think the `get_long_attr` function currently in the JSON implicitly expects nanosecond resolution, which would be incorrect and mangles the subsequent functions that convert to the unit of the encoder. I think the easiest approach would be to update the `get_long_attr` function to read the unit of the object and convert to nanosecond as required

**Comment by WillAyd** [comment]
Also wanted to add...this is awesome. Very impressive work

**Comment by jbrockmendel** [comment]
> update the get_long_attr function 

I tried adding

```
PyObject *nano_obj = PyObject_CallMethod(o, "_as_unit", "(s)", "ns");
```

and am getting a segfault.  is there a better way to do this?

**Comment by WillAyd** [comment]
I think you want

```python
PyObject *nano_obj = PyObject_CallMethod(o, "_as_unit", "s", "ns");
```

The `(s)` would mean you are building a tuple.

If you get stuck from there I would suggest running python with gdb attached to it via `gdb python` then doing something like `run -m pytest --lf` (or run whatever script gives you the segfault)

Assuming you built the extensions with debugging symbols enabled your debugger will drop you right at the point where the segfault occurs and let you inspect the state of things. If the above doesn't do it alone, my guess is there are other types of objects that we deal with than just timestamps in that function. Ideally we would do a type check...but we aren't at the moment, so dealing with anything but a Timestamp and trying to invoke that method could also cause a segfault

Lastly you might find this useful:

https://pandas.pydata.org/docs/dev/development/debugging_extensions.html#improve-debugger-printing

Good luck. If you get stuck ping me again and can look deeper

**Comment by jbrockmendel** [comment]
I ended up effectively re-implementing _as_unit.  No more segfault, but im just getting wrong answers.

```
static npy_int64 get_long_attr(PyObject *o, const char *attr) {
    // NB we are implicitly assuming that o is a Timedelta or Timestamp, or NaT

    npy_int64 long_val;
    PyObject *value = PyObject_GetAttrString(o, attr);
    long_val =
        (PyLong_Check(value) ? PyLong_AsLongLong(value) : PyLong_AsLong(value));

    Py_DECREF(value);

    if (long_val == NPY_MIN_INT64) {
        // i.e. o is NaT
        return long_val;
    }

    // ensure we are in nanoseconds, similar to Timestamp._as_reso or _as_unit
    NPY_DATETIMEUNIT reso = (NPY_DATETIMEUNIT)PyObject_GetAttrString(o, "_reso");

    if (reso == NPY_FR_us) {
        long_val = long_val * 1000L;
    } else if (reso == NPY_FR_ms) {
        long_val = long_val * 1000000L;
    } else if (reso == NPY_FR_s) {
        long_val = long_val * 1000000000L;
    }

    return long_val;
}
```

The test case thats failing _should_ go through the NPY_FR_us branch, but apparently none of these conditions are evaluating as truthy.

**Comment by WillAyd** [comment]
Ah I gotcha. The problem there is you essentially are getting back a `PyObject` - casting it to an NPY_DATETIMEUNIT doesn't change anything about the object itself or the bytes that comprise it; in this case you would just errantly be suppressing warnings with the cast

If you want the primitive integer value for comparison against the NPY_DATETIMEUNIT enum you'd likely want to do something like:

```c
if (!PyLong_Check(reso)) {
  // TODO: we should have error handling here, but one step at a time...
}

long cReso = PyLong_AsLong(reso);
if (cReso == -1 && PyErr_Occurred()) {
  // TODO: we should have error handling here, but one step at a time...
}
```

And then use `cReso` for your comparisons

Also be sure to `Py_DECREF(reso)` when you are done with it

**Comment by WillAyd** [comment]
@jbrockmendel just to drive home the issue with the cast, I think it is helpful to think about the bytes that are getting moved around. When comparing against the NPY_DATETIMEUNIT enum to see if something is in milliseconds, you are essentially comparisong against the integer 8. Assuming a 32 bit integer on a little ending platform, the binary structure of that is `00000000 00000000 00000000 00001000`. 

By contrast, a PyObject is a struct that will occupy more memory than 32 bits and have a totally different binary representation. It's not important what exactly that is for now, but let's assume that it's  `00110011 00110011 00110011 00110011 00110011 00110011 00110011 00110011` and so on...

When you cast the PyObject to a NPY_DATETIMEUNIT you are basically just casting it to an integer and therefore only interpreting an integer's amount of bytes. I don't know what the exact rules for that are and if its well defined, but let's assume for simplicity it just takes the first 32 bites of the PyObject. Your previous comparisons would then just be trying to compare against `00110011 00110011 00110011 00110011`, which doesn't equal the 8 you are looking for.

File: pandas/_libs/tslibs/conversion.pyx [hunk 1]
```diff
@@ -292,11 +298,19 @@ cdef _TSObject convert_to_tsobject(object ts, tzinfo tz, str unit,
...
+                # TODO: what if user explicitly passes nanos=0?
+                reso = NPY_FR_us
+        else:
+            reso = NPY_FR_ns
+        return convert_datetime_to_tsobject(ts, tz, nanos, reso=reso)
     elif PyDate_Check(ts):
         # Keep the converter same as PyDateTime's
         ts = datetime.combine(ts, time())
-        return convert_datetime_to_tsobject(ts, tz)
+        return convert_datetime_to_tsobject(ts, tz, nanos=0, reso=NPY_FR_us)  # TODO: or lower?
```
Comment (mroeschke) [inline pandas/_libs/tslibs/conversion.pyx:313]: IMO I would cast dates to the lowest resolution (and document it)

File: pandas/_libs/src/ujson/python/objToJSON.c [hunk 2]
```diff
@@ -278,11 +278,40 @@ static int is_simple_frame(PyObject *obj) {
...
+    // NB we are implicitly assuming that o is a Timedelta or Timestamp, or NaT
+
     npy_int64 long_val;
     PyObject *value = PyObject_GetAttrString(o, attr);
     long_val =
         (PyLong_Check(value) ? PyLong_AsLongLong(value) : PyLong_AsLong(value));
+
     Py_DECREF(value);
+
+    if (long_val == NPY_MIN_INT64) {
```
Comment (WillAyd) [inline pandas/_libs/src/ujson/python/objToJSON.c:290]: I guess this can be completely replaced by `object_is_nat_type(o)`

File: pandas/_libs/src/ujson/python/objToJSON.c [hunk 3]
```diff
@@ -278,11 +278,40 @@ static int is_simple_frame(PyObject *obj) {
...
+        // i.e. o is NaT
+        return long_val;
+    }
+
+    // ensure we are in nanoseconds, similar to Timestamp._as_creso or _as_unit
+    PyObject* reso = PyObject_GetAttrString(o, "_creso");
+    // if (!PyLong_Check(reso)) {
```
Comment (WillAyd) [inline pandas/_libs/src/ujson/python/objToJSON.c:297]:
```suggestion
    if (!PyLong_Check(reso))
      return -1;
```

Might as well set this up now. The caller should be checking for -1 and PyErr_Occurred() (can be a separate PR if not)

File: pandas/_libs/src/ujson/python/objToJSON.c [hunk 4]
```diff
@@ -278,11 +278,40 @@ static int is_simple_frame(PyObject *obj) {
...
+    // TODO(anyone): we should have error handling here, but one step at a time
+    // }
+
+    long cReso = PyLong_AsLong(reso);
+    // if (cReso == -1 && PyErr_Occurred()) {
```
Comment (WillAyd) [inline pandas/_libs/src/ujson/python/objToJSON.c:302]:
```suggestion
    if (cReso == -1 && PyErr_Occurred())
      return -1;
```

File: pandas/_libs/src/ujson/python/objToJSON.c [hunk 5]
```diff
@@ -278,11 +278,40 @@ static int is_simple_frame(PyObject *obj) {
...
+    // TODO(anyone): we should have error handling here, but one step at a time
+    // }
+    Py_DECREF(reso);
```
Comment (WillAyd) [inline pandas/_libs/src/ujson/python/objToJSON.c:305]: Should move this before the above check

File: pandas/_libs/tslib.pyx [hunk 6]
```diff
@@ -551,7 +551,7 @@ cpdef array_to_datetime(
                             raise ValueError('Cannot mix tz-aware with '
                                              'tz-naive values')
                         if isinstance(val, _Timestamp):
-                            iresult[i] = val.as_unit("ns").value
+                            iresult[i] = (<_Timestamp>val)._as_creso(NPY_FR_ns).value
```
Comment (WillAyd) [inline pandas/_libs/tslib.pyx:554]: Does `.value` naturally not return a Timestamp here?

File: pandas/_libs/tslibs/timestamps.pyx [hunk 7]
```diff
@@ -497,9 +497,9 @@ cdef class _Timestamp(ABCTimestamp):
             # Matching numpy, we cast to the higher resolution. Unlike numpy,
             #  we raise instead of silently overflowing during this casting.
             if self._creso < other._creso:
-                self = (<_Timestamp>self)._as_creso(other._creso, round_ok=False)
+                self = (<_Timestamp>self)._as_creso(other._creso, round_ok=True)
```
Comment (WillAyd) [inline pandas/_libs/tslibs/timestamps.pyx:500]: Similar question here. Generally pretty wary of casting in C / Cython without explicit type checks, though ideally the type system handles for us

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/timestamps.pyx (hunk 7 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/timestamps.pyx:500]: thats fair, but the cast is present ATM so id like to consider this out of scope

File: pandas/_libs/tslib.pyx (hunk 6 above)
Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx:554]: `.value` returns an `int64_t`

File: pandas/_libs/src/ujson/python/objToJSON.c [hunk 8]
```diff
@@ -278,11 +278,40 @@ static int is_simple_frame(PyObject *obj) {
 }
 
 static npy_int64 get_long_attr(PyObject *o, const char *attr) {
+    // NB we are implicitly assuming that o is a Timedelta or Timestamp, or NaT
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/_libs/src/ujson/python/objToJSON.c:281]: that would entail a bunch more boilerplate in the ujson.c file i think.  doing the validation on the ".value" attribute should be enough.

File: pandas/_libs/src/ujson/python/objToJSON.c (hunk 5 above)
Comment (jbrockmendel) [reply pandas/_libs/src/ujson/python/objToJSON.c:305]: should this also occur inside the short-circuiting return on reso validation?

Comment (WillAyd) [reply pandas/_libs/src/ujson/python/objToJSON.c:303]: Nope - you only need to DECREF reso one time. If you DECREF again in the return you would be double freeing.

The assignment to `cReso` copies the value that you want, so reso can be DECREFed immediately thereafter whether things failed or not

File: pandas/_libs/src/ujson/python/objToJSON.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (WillAyd) [reply pandas/_libs/src/ujson/python/objToJSON.c:281]: Where is that done? The problem with our C code is that we have too many assumptions around object types, so refactoring is a nightmare. When our assumptions don't hold true we get undefined behavior and segfaults

File: pandas/_libs/src/ujson/python/objToJSON.c (hunk 5 above)
Comment (jbrockmendel) [reply pandas/_libs/src/ujson/python/objToJSON.c:303]: i mean before L299.  if we go reach L299, won't this Py_DECREF not get hit?

File: pandas/_libs/src/ujson/python/objToJSON.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/_libs/src/ujson/python/objToJSON.c:281]: im not sure i understand the question.  The .value lookup is the PyObject_GetAttrString on L284.  And both the places where this is called from are after hasattr checks.

In the past weve discussed porting this module to cython piecemeal and that never really panned out.  Might be worth trying an all-at-once one of these days.

File: pandas/_libs/tslib.pyx (hunk 6 above)
Comment (WillAyd) [reply pandas/_libs/tslib.pyx:554]: Ah gotcha. The cast isn't needed though right? Even `pydatetime_to_dt64` returns an int64_t so probably better to remove this

Comment (jbrockmendel) [reply pandas/_libs/tslib.pyx:554]: ill just revert this, as the code currently there should work fine (just less performant)

File: pandas/_libs/src/ujson/python/objToJSON.c (hunk 5 above)
Comment (WillAyd) [reply pandas/_libs/src/ujson/python/objToJSON.c:303]: Ah good catch - yea you are correct

File: pandas/_libs/src/ujson/python/objToJSON.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (WillAyd) [reply pandas/_libs/src/ujson/python/objToJSON.c:281]: Ah OK I see your point - didn't notice that in the diff. The duck typed approach would be to make sure value isn't NULL on L284. Otherwise right now things probably segfault.

As far as porting to Cython...that would be a ton of work. Not sure that would make things easier either with the way ultrajson works

[the remark this answers is not in the document]
Comment (WillAyd) [reply pandas/_libs/src/ujson/python/objToJSON.c:281]: I opened #49756 to track this separately from this PR

File: pandas/_libs/src/ujson/python/objToJSON.c (hunk 2 above)
Comment (WillAyd) [reply pandas/_libs/src/ujson/python/objToJSON.c:290]: Can you still use this function? Shouldn't require the creation of anything new

File: pandas/_libs/tslibs/conversion.pyx [hunk 9]
```diff
@@ -285,11 +291,22 @@ cdef _TSObject convert_to_tsobject(object ts, tzinfo tz, str unit,
             obj.value = ts
             pandas_datetime_to_datetimestruct(ts, NPY_FR_ns, &obj.dts)
     elif PyDateTime_Check(ts):
-        return convert_datetime_to_tsobject(ts, tz, nanos)
+        if nanos == 0:
+            if isinstance(ts, ABCTimestamp):
+                reso = abbrev_to_npy_unit(ts.unit)  # TODO: faster way to do this?
+            else:
+                # TODO: what if user explicitly passes nanos=0?
```
Comment (WillAyd) [inline pandas/_libs/tslibs/conversion.pyx:298]: Is it possible to hit this? Maybe we should raise instead?

**Review by WillAyd — COMMENTED** [review]
Can you add a whatsnew for this too?

File: pandas/_libs/tslibs/conversion.pyx (hunk 9 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/conversion.pyx:298]: i think it could happen with `pd.Timestamp(pydatetime_obj, nanosecond=0)`

File: pandas/core/arrays/datetimes.py [hunk 10]
```diff
@@ -508,7 +508,10 @@ def _unbox_scalar(self, value) -> np.datetime64:
         if not isinstance(value, self._scalar_type) and value is not NaT:
             raise ValueError("'value' should be a Timestamp.")
         self._check_compatible_with(value)
-        return value.asm8
+        if value is NaT:
```
Comment (WillAyd) [inline pandas/core/arrays/datetimes.py:511]: Possible I've missed this conversation but do we need to give consideration to a generic NaT type that can hold different precisions? Or are we always going to use numpy's value?

File: pandas/core/array_algos/take.py [hunk 11]
```diff
@@ -360,7 +360,14 @@ def wrapper(
         if out_dtype is not None:
             out = out.view(out_dtype)
         if fill_wrap is not None:
+            # FIXME: if we get here with dt64/td64 we need to be sure we have
+            #  matching resos
+            if fill_value.dtype.kind == "m":
```
Comment (WillAyd) [inline pandas/core/array_algos/take.py:365]: Might be worth making a re-usable function for this? I could see this useful in other areas. Something like `reso_for_type`

**Review by WillAyd — APPROVED** [review]
I think generally lgtm. A few comments that aren't blockers

File: pandas/_libs/tslibs/conversion.pyx [hunk 12]
```diff
@@ -288,11 +294,22 @@ cdef _TSObject convert_to_tsobject(object ts, tzinfo tz, str unit,
...
+                reso = NPY_FR_us
+        else:
+            reso = NPY_FR_ns
+        return convert_datetime_to_tsobject(ts, tz, nanos, reso=reso)
     elif PyDate_Check(ts):
         # Keep the converter same as PyDateTime's
+        # For date object we give the lowest supporte resolution, ie. "s"
```
Comment (mroeschke) [inline pandas/_libs/tslibs/conversion.pyx:308]: Do we have a test where we construct a Timestamp from a datetime.date?

Comment (mroeschke) [inline pandas/_libs/tslibs/conversion.pyx:308]:
```suggestion
        # For date object we give the lowest supported resolution, i.e. "s"
```

**Review by mroeschke — APPROVED** [review]
LGTM can merge on green

File: pandas/_libs/tslibs/conversion.pyx (hunk 12 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/conversion.pyx]: dedicated test added + green

**Comment by mroeschke** [comment]
Thanks @jbrockmendel

**Comment by jbrockmendel** [comment]
thanks for reviews @mroeschke and @willayd.  we're getting close to having full non-nano support!
